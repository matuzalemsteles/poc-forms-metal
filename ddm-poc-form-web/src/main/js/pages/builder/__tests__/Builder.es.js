import Builder from '../Builder.es';

let component;
let spritemap = 'icons.svg';

describe('Builder', () => {
    beforeEach(() => {
        jest.useFakeTimers();
    });

    afterEach(() => {
        if (component) {
            component.dispose();
        }
    });

    it('should render the default markup', () => {
        component = new Builder({
            spritemap,
        });

        expect(component).toMatchSnapshot();
    });

    it('should continue to propagate the fieldAdd event', () => {
        component = new Builder({
            spritemap,
        });

        const spy = jest.spyOn(component, 'emit');
        const { sidebar } = component.refs;
        const mockEvent = jest.fn();

        sidebar.emit('fieldAdd', mockEvent);

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
            'fieldAdd',
            expect.anything()
        );
    });

    it('should continue to propagate the fieldEdit event', () => {
        component = new Builder({
            spritemap,
        });

        const spy = jest.spyOn(component, 'emit');
        const { sidebar } = component.refs;
        const mockEvent = jest.fn();

        sidebar.emit('fieldEdit', mockEvent);

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
            'fieldEdit',
            expect.anything()
        );
    });

    it('should continue to propagate the fieldMove event', () => {
        component = new Builder({
            spritemap,
        });

        const spy = jest.spyOn(component, 'emit');
        const { layoutRenderer } = component.refs;
        const mockEvent = jest.fn();

        layoutRenderer.emit('fieldMove', mockEvent);

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
            'fieldMove',
            expect.anything()
        );
    });

    it('should continue to propagate the fieldDelete event', () => {
        component = new Builder({
            spritemap,
        });

        const spy = jest.spyOn(component, 'emit');
        const { layoutRenderer } = component.refs;
        const mockEvent = jest.fn();

        layoutRenderer.emit('deleteButtonClicked', mockEvent);

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
            'fieldDelete',
            expect.anything()
        );
    });

    it('should continue to propagate the fieldClicked event and open the sidebar', () => {
        component = new Builder({
            spritemap,
        });

        const spy = jest.spyOn(component, 'emit');
        const { layoutRenderer, sidebar } = component.refs;
        const mockEvent = jest.fn();

        layoutRenderer.emit('fieldClicked', mockEvent);

        jest.runAllTimers();

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
            'fieldClicked',
            expect.anything()
        );
        expect(sidebar.state.show).toBeTruthy();
    });

    it('should open the sidebar when click the creation button', () => {
        component = new Builder({
            spritemap,
        });

        const { managementToolbar, sidebar } = component.refs;

        managementToolbar.refs.creationMenu.element.click();

        jest.runAllTimers();

        expect(sidebar.props.mode).toBe('add');
        expect(sidebar.state.mode).toBe('add');
        expect(sidebar.state.show).toBeTruthy();
    });
});