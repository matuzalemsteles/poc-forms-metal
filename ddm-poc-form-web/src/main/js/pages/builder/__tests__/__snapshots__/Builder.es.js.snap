// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder should render the default markup 1`] = `
<div>
  <div>
    <nav class="management-bar management-bar-light navbar navbar-expand-md">
      <div class="container-fluid container-fluid-max-xl">
        <ul class="navbar-nav"></ul>
        <ul class="navbar-nav">
          <li class="nav-item">
            <button class="btn nav-btn nav-btn-monospaced btn-primary" aria-label="plus" title="new" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
                <title>plus</title>
                <use xlink:href="icons.svg#plus"></use>
              </svg>
            </button>
          </li>
        </ul>
      </div>
    </nav>
  </div>
  <div class="container">
    <div class="sheet">
      <p>Drag from sidebar and drop here</p>
    </div>
  </div>
  <div class="sidebar-container" ref="sidebar">
    <div class="sidebar sidebar-light">
      <nav class="component-tbar tbar">
        <div class="container-fluid">
          <ul class="tbar-nav">
            <li class="tbar-item tbar-item-expand text-left">
              <div class="tbar-section">
                <span class="text-truncate-inline">
                  <span class="text-truncate">Add Elements</span>
                </span>
              </div>
            </li>
            <li class="tbar-item">
              <a class="component-action" href="#1" role="button" ref="close">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                  <use xlink:href="icons.svg#times"></use>
                </svg>
              </a>
            </li>
          </ul>
        </div>
      </nav>
      <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
        <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
          <span class="navbar-text-truncate">Details</span>
          <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
            <use xlink:href="icons.svg#caret-bottom"></use>
          </svg>
        </a>
        <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
          <ul class="nav navbar-nav" role="tablist">
            <li ref="tab0" class="nav-item">
              <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
                <span class="navbar-text-truncate">Elements</span>
              </a>
            </li>
          </ul>
        </div>
      </nav>
      <div class="ddm-sidebar-body"></div>
    </div>
  </div>
</div>
`;
